Support running RemoteRenderingBackendProxy objects per-thread.

mattwoodrow opened the pull request:
#### 7ae4b48cc67ab0e8e4dcc4e7bca877e8f97e6a1b
<pre>
Support running RemoteRenderingBackendProxy objects per-thread.
<a href="https://bugs.webkit.org/show_bug.cgi?id=248418">https://bugs.webkit.org/show_bug.cgi?id=248418</a>

Reviewed by Kimmo Kinnunen.

Changes StreamClientConnection to use an IPC::Connection::Client rather than MessageReceiver for providing
callbacks to the consumer.

Converts RemoteRenderingBackendProxy and RemoteGraphicsContextGLProxy to use their StreamClientConnection to receive
notifications about the GPU process crashing from their StreamClientConnections rather than their GPUProcessConnection.

Makes RemoteRenderingBackendProxy create its StreamClientConnection using the provided SerialFunctionDispatcher, so that
we can send and receive all messages from that.

* Source/WebKit/Platform/IPC/StreamClientConnection.cpp:
(IPC::StreamClientConnection::DedicatedConnectionClient::DedicatedConnectionClient):
(IPC::StreamClientConnection::DedicatedConnectionClient::didClose):
(IPC::StreamClientConnection::open):
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::RemoteGraphicsContextGLProxy):
(WebKit::RemoteGraphicsContextGLProxy::paintCompositedResultsToVideoFrame):
(WebKit::RemoteGraphicsContextGLProxy::didClose):
(WebKit::RemoteGraphicsContextGLProxy::abandonGpuProcess):
(WebKit::RemoteGraphicsContextGLProxy::disconnectGpuProcessIfNeeded):
(WebKit::RemoteGraphicsContextGLProxy::gpuProcessConnectionDidClose): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
(WebKit::RemoteGraphicsContextGLProxy::isContextLost const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::assertDispatcherIsCurrent const):
(WebKit::RemoteImageBufferProxy::hasPendingFlush const):
(WebKit::RemoteImageBufferProxy::didFlush):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::create):
(WebKit::RemoteRenderingBackendProxy::RemoteRenderingBackendProxy):
(WebKit::RemoteRenderingBackendProxy::~RemoteRenderingBackendProxy):
(WebKit::RemoteRenderingBackendProxy::ensureGPUProcessConnection):
(WebKit::RemoteRenderingBackendProxy::didClose):
(WebKit::RemoteRenderingBackendProxy::disconnectGPUProcess):
(WebKit::RemoteRenderingBackendProxy::finalizeRenderingUpdate):
(WebKit::RemoteRenderingBackendProxy::didPaintLayers):
(WebKit::RemoteRenderingBackendProxy::gpuProcessConnectionDidClose): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
(WebKit::RemoteRenderingBackendProxy::isGPUProcessConnectionClosed const):
(WebKit::RemoteRenderingBackendProxy::dispatcher):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.cpp:
(WebKit::RemoteGPUProxy::RemoteGPUProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/RemoteGraphicsContextGLProxyCocoa.mm:
(WebKit::RemoteGraphicsContextGLProxy::create):
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:

Canonical link: <a href="https://commits.webkit.org/257266@main">https://commits.webkit.org/257266@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bd851a8ce3a23cfa1020383a2ba3d6e2686215a8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98070 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7287 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31223 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107533 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167801 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102010 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7774 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/36050 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90677 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104145 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103722 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5849 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84670 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32978 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87718 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89432 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75915 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1261 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/20866 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1226 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22369 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5006 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6086 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/44819 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2526 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41774 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:
```diff
@@ -99,6 +99,9 @@ class MockMessageReceiver : public IPC::MessageReceiver, public WaitForMessageMi
         return false;
     }
+
+    void didClose(IPC::Connection&) final { }
```
Would it be possible to add tests for delivering the didclose?

E.g. server does invalidate() and the client catches didClose?

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.cpp:
```diff
@@ -45,31 +46,30 @@ static constexpr size_t defaultStreamSize = 1 << 21;
 {
     m_streamConnection->invalidate();
-    m_gpuProcessConnection = nullptr;
+    m_streamConnection = nullptr;
```
I think you cannot assign like this as the other webgpu objects might use the streamConnection() to obtain the connection and send to it.
Sending to invalid/closed connection is ok.
Currently we don't have a  sequence where we would guarantee no RemoteGPU**Proxy objects be accessed after losing the context. So it might be simpler to leave the stream connection ptr as is, for now.

mattwoodrow replied:
I'll move these changes to a separate issue, but do you think the solution here is to keep m_streamConnection, and implementing WebGL's `isContextLost()` using `m_streamConnection->isValid()` (which doesn't exist, but could).

kkinnunen-apple replied:
It perhaps could, but care should be taken to figure out the semantics. Naive/expected/most probable implementation of  `SCC::isValid()` would flip immediately when we know the other process crashes, but currently `RGCGLP::isContextLost()` will flip only due to the notification is being posted.

E.g. there could be a logical ordering bugs that then for example early-out omit sending the forceContextLost() upwards the inheritance chain

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
```diff
@@ -65,17 +65,13 @@ namespace WebKit {
 class RemoteRenderingBackendProxy
-    : private IPC::MessageReceiver
-    , public GPUProcessConnection::Client {
+    : public IPC::Connection::Client {
```
maybe private..

mattwoodrow replied:
We need this to be public so that WeakPtr<RemoteRenderingBackendProxy> can be created (previously CanMakeWeakPtr was available through the public GPUProcessConnection::Client base).

kkinnunen-apple reviewed (commented):
I think it might be better to do this in this patch only for RRB and not WebGL / WebGPU, as there's the issue that the caller model of use does not take into account when the GPU Process goes away. This means that we might get calls into msg sending functions after GPU Process goes away. This is currently expected, as the message sends are just business as usual, ignored.
So after this is working for RRBP, we could modify webgl/webgpu in a separate patch?
I know you would have to add a empty didClose function and a fixme, but maybe that's simpler for now?

kkinnunen-apple reviewed (approved):
I think going forward we need to ensure that the async functions don't cause problems. I would think we need a new, thread-safe approach for starting up the process.

mattwoodrow committed:
Support running RemoteRenderingBackendProxy objects per-thread.
https://bugs.webkit.org/show_bug.cgi?id=248418

Reviewed by Kimmo Kinnunen.

Changes StreamClientConnection to use an IPC::Connection::Client rather than MessageReceiver for providing
callbacks to the consumer.

Converts RemoteRenderingBackendProxy and RemoteGraphicsContextGLProxy to use their StreamClientConnection to receive
notifications about the GPU process crashing from their StreamClientConnections rather than their GPUProcessConnection.

Makes RemoteRenderingBackendProxy create its StreamClientConnection using the provided SerialFunctionDispatcher, so that
we can send and receive all messages from that.

* Source/WebKit/Platform/IPC/StreamClientConnection.cpp:
(IPC::StreamClientConnection::DedicatedConnectionClient::DedicatedConnectionClient):
(IPC::StreamClientConnection::DedicatedConnectionClient::didClose):
(IPC::StreamClientConnection::open):
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::RemoteGraphicsContextGLProxy):
(WebKit::RemoteGraphicsContextGLProxy::paintCompositedResultsToVideoFrame):
(WebKit::RemoteGraphicsContextGLProxy::didClose):
(WebKit::RemoteGraphicsContextGLProxy::abandonGpuProcess):
(WebKit::RemoteGraphicsContextGLProxy::disconnectGpuProcessIfNeeded):
(WebKit::RemoteGraphicsContextGLProxy::gpuProcessConnectionDidClose): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
(WebKit::RemoteGraphicsContextGLProxy::isContextLost const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::assertDispatcherIsCurrent const):
(WebKit::RemoteImageBufferProxy::hasPendingFlush const):
(WebKit::RemoteImageBufferProxy::didFlush):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::create):
(WebKit::RemoteRenderingBackendProxy::RemoteRenderingBackendProxy):
(WebKit::RemoteRenderingBackendProxy::~RemoteRenderingBackendProxy):
(WebKit::RemoteRenderingBackendProxy::ensureGPUProcessConnection):
(WebKit::RemoteRenderingBackendProxy::didClose):
(WebKit::RemoteRenderingBackendProxy::disconnectGPUProcess):
(WebKit::RemoteRenderingBackendProxy::finalizeRenderingUpdate):
(WebKit::RemoteRenderingBackendProxy::didPaintLayers):
(WebKit::RemoteRenderingBackendProxy::gpuProcessConnectionDidClose): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
(WebKit::RemoteRenderingBackendProxy::isGPUProcessConnectionClosed const):
(WebKit::RemoteRenderingBackendProxy::dispatcher):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.cpp:
(WebKit::RemoteGPUProxy::RemoteGPUProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/RemoteGraphicsContextGLProxyCocoa.mm:
(WebKit::RemoteGraphicsContextGLProxy::create):
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:

Canonical link: https://commits.webkit.org/257266@main
